src/historywork/DownloadBucketsWork.cpp

```diff
@@ -40,41 +45,62 @@ DownloadBucketsWork::getStatus() const
 {
     if (mState == WORK_RUNNING || mState == WORK_PENDING)
     {
-        return "downloading buckets";
+        if (!mHashes.empty())
+        {
+            auto numDone = std::distance(mHashes.begin(), mNextBucketIndex);
+            auto total = static_cast<uint32_t>(mHashes.size());
+            auto pct = (100 * numDone) / total;
+            return fmt::format(
+                "downloading and verifying buckets: {:d}/{:d} ({:d}%)", numDone,
+                total, pct);
+        }
     }
     return Work::getStatus();
 }
 
+bool
+DownloadBucketsWork::hasNext()
+{
+    return mNextBucketIndex != mHashes.end();
+}
+
 void
-DownloadBucketsWork::onReset()
+DownloadBucketsWork::resetIter()
 {
-    clearChildren();
+    mNextBucketIndex = mHashes.begin();
+}
 
-    for (auto const& hash : mHashes)
+std::string
+DownloadBucketsWork::yieldMoreWork()
+{
+    if (!hasNext())
     {
-        FileTransferInfo ft(mDownloadDir, HISTORY_FILE_TYPE_BUCKET, hash);
-        // Each bucket gets its own work-chain of
-        // download->gunzip->verify
-
-        auto verify = addWork<VerifyBucketWork>(mBuckets, ft.localPath_nogz(),
-                                                hexToBin256(hash));
-        verify->addWork<GetAndUnzipRemoteFileWork>(ft);
-        mDownloadBucketStart.Mark();
+        throw std::runtime_error("Nothing to iterate over!");
     }
+
+    auto hash = *mNextBucketIndex;
+    FileTransferInfo ft(mDownloadDir, HISTORY_FILE_TYPE_BUCKET, hash);
+    auto verify = addWork<VerifyBucketWork>(mBuckets, ft.localPath_nogz(),
+                                            hexToBin256(hash));
+    auto download = verify->addWork<GetAndUnzipRemoteFileWork>(ft);
+
+    assert(mVerifyDownloadMap.find(verify->getUniqueName()) ==
+           mVerifyDownloadMap.end());
+    mVerifyDownloadMap.insert(
+        std::make_pair(verify->getUniqueName(), download));
+    mDownloadBucketStart.Mark();
+
+    ++mNextBucketIndex;
+    return verify->getUniqueName();
 }
 
 void
 DownloadBucketsWork::notify(std::string const& child)
 {
-    auto i = mChildren.find(child);
-    if (i == mChildren.end())
-    {
-        CLOG(WARNING, "Work")
-            << "DownloadBucketsWork notified by unknown child " << child;
-        return;
-    }
+    auto downloadWork = mVerifyDownloadMap.find(child);
```

Comment:
Looks like `mVerifyDownloadMap` is only used to ensure that we don't update the wrong metrics - I would ditch `mVerifyDownloadMap` entirely and keep the code simple.
If there is a bug in how notify is called we have bigger problems as in the code base we seem to ignore those. On that front, (and general), we should log `notified by unknown child` with an `ERROR` log level instead of `INFO`.